src/adapter/src/catalog.rs

```diff
@@ -5711,6 +5773,154 @@ impl Catalog {
                         }),
                     )?;
                 }
+                Op::GrantPrivilege {
+                    object_id,
+                    privilege,
+                } => match object_id {
+                    ObjectId::Cluster(id) => {
+                        let cluster_name = state.get_cluster(id).name().to_string();
+                        if id.is_system() {
+                            return Err(AdapterError::Catalog(Error::new(
+                                ErrorKind::ReadOnlyCluster(cluster_name),
+                            )));
+                        }
+                        builtin_table_updates.push(state.pack_cluster_update(&cluster_name, -1));
+                        let cluster = state.get_cluster_mut(id);
+                        Self::grant_object_privilege(&mut cluster.privileges, privilege);
+                        tx.update_cluster(id, cluster)?;
+                        builtin_table_updates.push(state.pack_cluster_update(&cluster_name, 1));
+                    }
+                    ObjectId::Database(id) => {
+                        let database = state.get_database(&id);
```

Comment:
Why do database and schema not have a id.is_system check?

Reply:
That's an oversight, I've updated it to include those checks.